src/payment/strategies/affirm/affirm-payment-strategy.ts

```diff
@@ -0,0 +1,283 @@
+import { OrderActionCreator, OrderRequestBody } from '../../../order';
+import { OrderFinalizationNotRequiredError } from '../../../order/errors';
+import { PaymentArgumentInvalidError, PaymentMethodCancelledError } from '../../errors';
+import PaymentActionCreator from '../../payment-action-creator';
+import PaymentMethodActionCreator from '../../payment-method-action-creator';
+import { PaymentInitializeOptions, PaymentRequestOptions } from '../../payment-request-options';
+import PaymentStrategy from '../payment-strategy';
+
+import { Affirm, AffirmAddress, AffirmDiscount, AffirmItem, AffirmRequestData, SuccessAffirm } from './affirm';
+import AffirmScriptLoader from './affirm-script-loader';
+
+export default class AffirmPaymentStrategy implements PaymentStrategy {
+
+    private affirm?: Affirm;
+
+    constructor(
+        private _store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _affirmScriptLoader: AffirmScriptLoader
+    ) { }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        return this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(options.methodId))
+            .then(state => {
+                const paymentMethod = state.paymentMethods.getPaymentMethod(options.methodId);
+
+                if (!paymentMethod || !paymentMethod.clientToken) {
+                    throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+                }
+
+                const { config: { testMode }, clientToken: publicKey } = paymentMethod;
+
+                return this._affirmScriptLoader.load(publicKey, testMode);
+
+            }).then(affirm => {
+                this.affirm = affirm;
+
+                return this._store.getState();
+            });
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const paymentId = payload.payment && payload.payment.methodId;
+        const { useStoreCredit } = payload;
+        const { affirm } = this;
+
+        if (!affirm) {
+            throw new NotInitializedError(NotInitializedErrorType.PaymentNotInitialized);
+        }
+
+        if (!paymentId) {
+            throw new PaymentArgumentInvalidError(['payment.methodId']);
+        }
+
+        return this._store.dispatch(this._orderActionCreator.submitOrder({ useStoreCredit }, options))
+            .then((): Promise<SuccessAffirm> => {
+                affirm.checkout(this._getCheckoutInformation(useStoreCredit));
+
+                return new Promise((resolve, reject) => {
+                    affirm.checkout.open({
+                        onFail: () => { reject(new PaymentMethodCancelledError()); },
```

Comment:
I don't think all errors are cancellation error. So you might need to do a check here. Cancellation errors are useful when we want to distinguish between real errors and "errors" as a result of intentional cancellation by the user.